platform-core/src/main/java/com/webank/wecube/platform/core/service/cmder/ssh2/impl/PooledRemoteCommandExecutor.java

```diff
@@ -0,0 +1,141 @@
+
+import ch.ethz.ssh2.ChannelCondition;
+import ch.ethz.ssh2.Connection;
+import ch.ethz.ssh2.Session;
+import ch.ethz.ssh2.StreamGobbler;
+
+public class PooledRemoteCommandExecutor implements PoolableRemoteCommandExecutor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PooledRemoteCommandExecutor.class);
+    private Session sshSession;
+    private Connection sshConn;
+    private ExecutorService executor;
+    private RemoteCommandExecutorConfig config;
+
+    @Override
+    public String execute(RemoteCommand cmd) {
+        Future<String> f = this.executor.submit(new PooledRemoteCommandTask(cmd));
+
+        try {
+            return f.get();
+        } catch (InterruptedException | ExecutionException e1) {
+            LOGGER.error("execution error", e1);
+            return null;
+        }
+    }
+
+    @Override
+    public void execute(RemoteCommand cmd, boolean asynchorously) {
+        this.executor.submit(new PooledRemoteCommandTask(cmd));
+    }
+
+    @Override
+    public void init(RemoteCommandExecutorConfig config) throws Exception {
+        this.executor = Executors.newFixedThreadPool(3);
+        this.config = config;
+
+        try {
+            buildConnection();
+            buildSession();
+        } catch (IOException e) {
+            LOGGER.error("failed to init executor", e);
+            this.destroy();
+            throw new Exception("failed to init executor");
+        }
+    }
+
+    @Override
+    public void destroy() {
+        if (this.executor != null) {
+            this.executor.shutdown();
+        }
+
+        if (this.sshSession != null) {
+            this.sshSession.close();
+        }
+
+        if (this.sshConn != null) {
+            this.sshConn.close();
+        }
+    }
+
+    protected void buildConnection() throws IOException {
+        Connection conn = new Connection(getConfig().getRemoteHost(), getConfig().getPort());
+        conn.connect();
+        boolean isAuthenticated = conn.authenticateWithPassword(getConfig().getUser(), getConfig().getPsword());
+        if (!isAuthenticated) {
+            if (conn != null) {
+                conn.close();
+            }
+
+            throw new IOException("authentication failed");
+        }
+
+        this.sshConn = conn;
+    }
+
+    protected RemoteCommandExecutorConfig getConfig() {
+        return this.config;
+    }
+
+    protected void buildSession() throws IOException {
+        this.sshSession = this.sshConn.openSession();
+    }
+
+    private Session getSession() {
+        return sshSession;
+    }
+
+    private class PooledRemoteCommandTask implements Callable<String> {
+        private RemoteCommand cmd;
+
+        public PooledRemoteCommandTask(RemoteCommand cmd) {
+            super();
+            this.cmd = cmd;
+        }
+
+        @Override
+        public String call() throws Exception {
+            try {
+                getSession().execCommand(cmd.getCommand());
```

Comment:
need a mechanism here to recreate the session if it's not valid anymore. otherwise, the call will always throw exception once something wrong with the conection.

Reply:
同上